File: packages/components/nodes/tools/TradingBacktest/tradingBacktestCLI.js
#!/usr/bin/env node

/**
 * Trading Backtest CLI - Production Grade
 * Standalone script to backtest trading strategies
 * 
 * Features:
 * - Comprehensive input validation
 * - Graceful error handling
 * - Helpful error messages
 * - Progress indicators
 * 
 * Usage:
 *   node tradingBacktestCLI.js --symbol BTC --start 2023-01-01 --end 2023-12-31 --min-profit 0.2
 */

const { BacktestEngine } = require('./backtestEngine')
const path = require('path')

/**
 * Main CLI function
 */
async function main() {
    const args = parseArgs(process.argv.slice(2))

    // Show help if no arguments or help flag
    if (args.help || Object.keys(args).length === 0) {
        showHelp()
        process.exit(0)
    }

    // Validate required arguments
    const validation = validateArgs(args)
    if (!validation.valid) {
        console.error('Error:', validation.error)
        console.error('')
        showHelp()
        process.exit(1)
    }

    const dataDir = args['data-dir'] || path.join(process.cwd(), 'trading_data')
    const minProfit = parseFloat(args['min-profit'] || '0.2')

    // Validate minProfit is a valid number
    if (isNaN(minProfit) || !isFinite(minProfit)) {
        console.error('Error: min-profit must be a valid number')
        process.exit(1)
    }

    printHeader(args, dataDir, minProfit)

    try {
        const engine = new BacktestEngine(dataDir, minProfit)

        console.log('Running multi-level backtest...')
        console.log('Level 1: Single indicator strategies...')
        console.log('Level 2: Combined indicator strategies...')
        console.log('Level 3: Dynamic indicator strategies...')
        console.log('Level 4: Neural network approach...')
        console.log('')

        const result = await engine.runBacktest({
            symbol: args.symbol,
            startDate: args.start,
            endDate: args.end,
            minProfitLossPercent: minProfit
        })

        printResults(result)
        process.exit(0)
        
    } catch (error) {
        console.error('')
        console.error('='.repeat(60))
        console.error('❌ BACKTEST FAILED')
        console.error('='.repeat(60))
        console.error('')
        console.error('Error:', error.message)
        console.error('')
        
        // Provide helpful suggestions based on error type
        if (error.message.includes('Invalid symbol')) {
            console.error('Suggestion: Use only letters, numbers, hyphens, and underscores')
            console.error('Example: BTC, ETH, AAPL, MSFT-USD')
        } else if (error.message.includes('Invalid date') || error.message.includes('date range')) {
            console.error('Suggestion: Ensure dates are in YYYY-MM-DD format')
            console.error('Example: --start 2023-01-01 --end 2023-12-31')
        } else if (error.message.includes('Insufficient data')) {
            console.error('Suggestion: Try a longer date range or different symbol')
        } else if (error.message.includes('data directory')) {
            console.error('Suggestion: Ensure the data directory is writable')
            console.error('Check permissions or specify a different directory with --data-dir')
        }
        
        console.error('')
        process.exit(1)
    }
}

/**
 * Validate command line arguments
 * @param {Object} args - Parsed arguments
 * @returns {Object} Validation result with {valid: boolean, error: string}
 */
function validateArgs(args) {
    if (!args.symbol) {
        return { valid: false, error: 'Missing required argument: --symbol' }
    }
    
    if (!args.start) {
        return { valid: false, error: 'Missing required argument: --start' }
    }
    
    if (!args.end) {
        return { valid: false, error: 'Missing required argument: --end' }
    }
    
    // Basic date format validation
    const dateRegex = /^\d{4}-\d{2}-\d{2}$/
    if (!dateRegex.test(args.start)) {
        return { valid: false, error: 'Invalid start date format. Use YYYY-MM-DD' }
    }
    
    if (!dateRegex.test(args.end)) {
        return { valid: false, error: 'Invalid end date format. Use YYYY-MM-DD' }
    }
    
    // Validate date values
    const startDate = new Date(args.start)
    const endDate = new Date(args.end)
    
    if (isNaN(startDate.getTime())) {
        return { valid: false, error: `Invalid start date: ${args.start}` }
    }
    
    if (isNaN(endDate.getTime())) {
        return { valid: false, error: `Invalid end date: ${args.end}` }
    }
    
    if (startDate >= endDate) {
        return { valid: false, error: 'Start date must be before end date' }
    }
    
    return { valid: true }
}

/**
 * Print header with configuration
 */
function printHeader(args, dataDir, minProfit) {
    console.log('='.repeat(60))
    console.log('Trading Strategy Backtest')
    console.log('='.repeat(60))
    console.log(`Symbol: ${args.symbol}`)
    console.log(`Start Date: ${args.start}`)
    console.log(`End Date: ${args.end}`)
    console.log(`Minimum P/L: ${minProfit}%`)
    console.log(`Data Directory: ${dataDir}`)
    console.log('='.repeat(60))
    console.log('')
}

/**
 * Print backtest results
 */
function printResults(result) {
    console.log('='.repeat(60))
    console.log('BACKTEST RESULTS')
    console.log('='.repeat(60))
    console.log('')

    if (!result || result.level === 0) {
        console.log('❌ No profitable strategy found with the given criteria.')
        console.log('')
        console.log('Try:')
        console.log('  - Adjusting the date range')
        console.log('  - Lowering the minimum profit requirement')
        console.log('  - Testing a different symbol')
    } else {
        console.log(`✅ Most Effective Strategy Found (Level ${result.level})`)
        console.log('')
        console.log(`Strategy: ${result.strategy}`)
        console.log(`Profit/Loss: ${safeToFixed(result.profitLossPercent, 2)}%`)
        console.log(`Win Rate: ${safeToFixed(result.winRate, 2)}%`)
        console.log(`Total Trades: ${result.totalTrades}`)
        console.log('')
        
        if (result.indicators && result.indicators.length > 0) {
            console.log('Indicators:')
            for (const indicator of result.indicators) {
                if (indicator && indicator.name && indicator.params) {
                    const params = Object.entries(indicator.params)
                        .map(([k, v]) => `${k}=${v}`)
                        .join(', ')
                    console.log(`  • ${indicator.name}(${params})`)
                }
            }
        }
    }

    console.log('='.repeat(60))
}

/**
 * Safely format number to fixed decimal places
 */
function safeToFixed(value, decimals) {
    if (typeof value !== 'number' || !isFinite(value)) {
        return '0.00'
    }
    try {
        return value.toFixed(decimals)
    } catch (error) {
        return '0.00'
    }
}

/**
 * Show help message
 */
function showHelp() {
    console.log('Usage: node tradingBacktestCLI.js --symbol <SYMBOL> --start <START_DATE> --end <END_DATE> [OPTIONS]')
    console.log('')
    console.log('Required Arguments:')
    console.log('  --symbol       Trading symbol (e.g., BTC, ETH, AAPL, TSLA)')
    console.log('  --start        Start date in YYYY-MM-DD format')
    console.log('  --end          End date in YYYY-MM-DD format')
    console.log('')
    console.log('Optional Arguments:')
    console.log('  --min-profit   Minimum profit/loss percentage (default: 0.2)')
    console.log('  --data-dir     Directory to store historical data (default: ./trading_data)')
    console.log('  --help         Show this help message')
    console.log('')
    console.log('Examples:')
    console.log('  node tradingBacktestCLI.js --symbol BTC --start 2023-01-01 --end 2023-12-31')
    console.log('  node tradingBacktestCLI.js --symbol ETH --start 2023-06-01 --end 2023-09-30 --min-profit 0.5')
    console.log('  node tradingBacktestCLI.js --symbol AAPL --start 2023-01-01 --end 2023-12-31 --data-dir /tmp/data')
    console.log('')
}

/**
 * Parse command line arguments
 * @param {Array<string>} args - Command line arguments
 * @returns {Object} Parsed arguments
 */
function parseArgs(args) {
    const result = {}
    
    for (let i = 0; i < args.length; i++) {
        if (args[i].startsWith('--')) {
            const key = args[i].substring(2)
            
            // Check for flags without values
            if (key === 'help' || key === 'version') {
                result[key] = true
                continue
            }
            
            const value = args[i + 1]
            if (value && !value.startsWith('--')) {
                result[key] = value
                i++
            } else {
                console.warn(`Warning: No value provided for --${key}`)
            }
        }
    }
    
    return result
}

// Run if executed directly
if (require.main === module) {
    main().catch((error) => {
        console.error('')
        console.error('Unexpected error:', error.message)
        console.error('')
        if (error.stack) {
            console.error('Stack trace:')
            console.error(error.stack)
        }
        process.exit(1)
    })
}

module.exports = { main }

File: packages/components/nodes/tools/TradingBacktest/tradingBacktestCLI.py
#!/usr/bin/env python3
"""
Trading Backtest CLI - Pure Python
Standalone script to backtest trading strategies

Usage:
  python tradingBacktestCLI.py --symbol BTC --start 2023-01-01 --end 2023-12-31 --min-profit 0.2
"""

import sys
import argparse
from backtestEngine import BacktestEngine


def main():
    """Main CLI function"""
    parser = argparse.ArgumentParser(
        description='Trading Strategy Backtest',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''
Examples:
  %(prog)s --symbol BTC --start 2023-01-01 --end 2023-12-31
  %(prog)s --symbol ETH --start 2023-06-01 --end 2023-09-30 --min-profit 0.5
  %(prog)s --symbol AAPL --start 2023-01-01 --end 2023-12-31 --data-dir /tmp/data
        '''
    )
    
    parser.add_argument('--symbol', required=True,
                       help='Trading symbol (e.g., BTC, ETH, AAPL, TSLA)')
    parser.add_argument('--start', required=True,
                       help='Start date in YYYY-MM-DD format')
    parser.add_argument('--end', required=True,
                       help='End date in YYYY-MM-DD format')
    parser.add_argument('--min-profit', type=float, default=0.2,
                       help='Minimum profit/loss percentage (default: 0.2)')
    parser.add_argument('--data-dir', default='./trading_data',
                       help='Directory to store historical data (default: ./trading_data)')
    
    args = parser.parse_args()
    
    print('=' * 60)
    print('Trading Strategy Backtest')
    print('=' * 60)
    print(f'Symbol: {args.symbol}')
    print(f'Start Date: {args.start}')
    print(f'End Date: {args.end}')
    print(f'Minimum P/L: {args.min_profit}%')
    print(f'Data Directory: {args.data_dir}')
    print('=' * 60)
    print('')
    
    try:
        engine = BacktestEngine(args.data_dir, args.min_profit)
        
        print('Running multi-level backtest...')
        print('Level 1: Single indicator strategies...')
        print('Level 2: Combined indicator strategies...')
        print('Level 3: Dynamic indicator strategies...')
        print('Level 4: Neural network approach...')
        print('')
        
        result = engine.run_backtest({
            'symbol': args.symbol,
            'startDate': args.start,
            'endDate': args.end,
            'minProfitLossPercent': args.min_profit
        })
        
        print('=' * 60)
        print('BACKTEST RESULTS')
        print('=' * 60)
        print('')
        
        if result['level'] == 0:
            print('❌ No profitable strategy found with the given criteria.')
            print('')
            print('Try:')
            print('  - Adjusting the date range')
            print('  - Lowering the minimum profit requirement')
            print('  - Testing a different symbol')
        else:
            print(f"✅ Most Effective Strategy Found (Level {result['level']})")
            print('')
            print(f"Strategy: {result['strategy']}")
            print(f"Profit/Loss: {result['profitLossPercent']:.2f}%")
            print(f"Win Rate: {result['winRate']:.2f}%")
            print(f"Total Trades: {result['totalTrades']}")
            print('')
            
            if result['indicators']:
                print('Indicators:')
                for indicator in result['indicators']:
                    params = ', '.join(f"{k}={v}" for k, v in indicator['params'].items())
                    print(f"  • {indicator['name']}({params})")
        
        print('=' * 60)
        return 0
        
    except ValueError as e:
        print('')
        print('=' * 60)
        print('❌ BACKTEST FAILED')
        print('=' * 60)
        print('')
        print(f'Error: {e}')
        print('')
        
        # Provide helpful suggestions
        error_msg = str(e)
        if 'Invalid symbol' in error_msg:
            print('Suggestion: Use only letters, numbers, hyphens, and underscores')
            print('Example: BTC, ETH, AAPL, MSFT-USD')
        elif 'Invalid date' in error_msg or 'date range' in error_msg:
            print('Suggestion: Ensure dates are in YYYY-MM-DD format')
            print('Example: --start 2023-01-01 --end 2023-12-31')
        elif 'Insufficient data' in error_msg:
            print('Suggestion: Try a longer date range or different symbol')
        elif 'data directory' in error_msg:
            print('Suggestion: Ensure the data directory is writable')
            print('Check permissions or specify a different directory with --data-dir')
        
        print('')
        return 1
    except Exception as e:
        print('')
        print(f'Unexpected error: {e}')
        print('')
        import traceback
        traceback.print_exc()
        return 1


if __name__ == '__main__':
    sys.exit(main())
